src/rs/lib.rs

```diff
@@ -950,30 +1007,64 @@ impl Default for Connection {
     }
 }
 
+// Remarks on Rust callback in general:
+// Callback function or closure is set to msquic handle context of type `*mut Box<TCallback>>`.
+// We cannot use type `*mut TCallback` because dyn trait cannot be converted to a pointer directly.
+// In the ffi function passed to msquic, the TCallback is extracted from msquic callback context
+// and invoked.
+// The context is cleaned up when connection is dropped or when new callback is set.
+// TCallback type is either Fn or FnMut type, so user can capture variables in them without worrying
+// about lifetime and how to set or cleanup msquic callback context.
+
+/// Connection callback type.
+/// msquic never invokes the connection callback in parallel, so use FnMut to allow mutation.
+type ConnectionCallback = dyn FnMut(ConnectionRef, ConnectionEvent) -> Result<(), Status> + 'static;
+
+extern "C" fn raw_conn_callback(
+    connection: HQUIC,
+    context: *mut c_void,
+    event: *mut ffi::QUIC_CONNECTION_EVENT,
+) -> QUIC_STATUS {
+    let conn = unsafe { ConnectionRef::from_raw(connection) };
+    let f = unsafe {
+        (context as *mut Box<ConnectionCallback>)
+            .as_mut() // allow mutation
+            .expect("cannot get ConnectionCallback from ctx")
+    };
+    let event = ConnectionEvent::from(unsafe { event.as_ref().unwrap() });
+    match f(conn, event) {
+        Ok(_) => StatusCode::QUIC_STATUS_SUCCESS.into(),
+        Err(e) => e.0,
+    }
+}
+
 impl Connection {
     pub fn new() -> Connection {
         Connection {
             handle: ptr::null_mut(),
         }
     }
 
-    /// TODO: The handler type should eventually be changed to Fn type.
-    /// ffi type and the context ptr makes this function unsafe.
-    pub fn open(
-        &mut self,
-        registration: &Registration,
-        handler: ffi::QUIC_CONNECTION_CALLBACK_HANDLER,
-        context: *const c_void,
-    ) -> Result<(), Status> {
+    pub fn open<F>(&mut self, registration: &Registration, handler: F) -> Result<(), Status>
+    where
+        F: FnMut(ConnectionRef, ConnectionEvent) -> Result<(), Status> + 'static,
```

Comment:
Isn't the lifetime of the connection enough here? Callbacks will likely have a context with a limited lifetime.

Reply:
Right, but isn't `'static` here requiring that the handler has a static lifetime? Why not use the lifetime of `self`?

Reply:
Ideally it should be lifetime of self. But Box<ConnectionCallback> cannot be annotated with lifetime. Maybe it is possible somehow or maybe more unsafe trick is needed. It can be a future improvement.